File: binary_search/disjoint_intervals/disjoint_intervals.py
from bisect import bisect_right


class RangesSummary:  # LeetCode Q.352.
    """Given integers stream, summarize all numbers seen so far as list of disjoint intervals."""

    def __init__(self):
        self.intervals = []  # Format: [inclusive left, inclusive right].

    def add_value(self, value: int):
        left, right = value, value  # Added value itself is a new range.

        idx = bisect_right(self.intervals, [left, left])  # Idx of the first range w/ start > left.
        if idx > 0 and self.intervals[idx - 1][1] + 1 >= left:  # Merge prev range.
            idx -= 1  # Insertion idx changes.
            right = max(right, self.intervals[idx][1])
            self.intervals[idx][1] = right

        else:
            self.intervals.insert(idx, [left, right])

        while idx + 1 < len(self.intervals) and self.intervals[idx + 1][0] - 1 <= right:
            right = max(right, self.intervals[idx + 1][1])
            self.intervals[idx][1] = right
            self.intervals.pop(idx + 1)  # Next range is merged.

    def get_intervals(self):
        return self.intervals

File: binary_search/disjoint_intervals/disjoint_intervals.cpp
#include <vector>
using namespace std;

class RangesSummary
{ // LeetCode Q.352.
private:
    vector<vector<int>> ranges; // Format: {inclusive left, inclusive right}.

public:
    RangesSummary() {}

    void add_value(int value)
    {
        int left = value, right = value; // Added value itself is a new range.

        int idx = upper_bound(
                      ranges.begin(), ranges.end(), left,
                      [](int val, const vector<int> &v)
                      { return val < v[0]; }) -
                  ranges.begin(); // Idx of the first range w/ start > left.

        if (idx > 0 && ranges[idx - 1][1] + 1 >= left)
        {             // Merge prev range.
            idx -= 1; // Insertion idx changes.
            right = max(right, ranges[idx][1]);
            ranges[idx][1] = right;
        }
        else
        {
            ranges.insert(ranges.begin() + idx, {left, right});
        }

        while (idx + 1 < ranges.size() && ranges[idx + 1][0] - 1 <= right)
        {
            right = max(right, ranges[idx + 1][1]);
            ranges[idx][1] = right;
            ranges.erase(ranges.begin() + idx + 1); // Next range is merged.
        }
    }

    vector<vector<int>> get_intervals()
    {
        return ranges;
    }
};